### ISSUE TITLE
Undefined behaviour with taking the address of a dereferenced nullptr in HalfedgeDS_list operator=

### ISSUE THREAD

**Issue by matthewlai** [CGAL/cgal#5728]
## Issue Details

In HalfedgeDS_list::pointer_update(), h_map[Halfedge_const_iterator()] calls Unique_hash_map::operator\[\](), which calls HashFunctor::operator(), which calls the iterator's operator*() in:
`
      operator()(const H& h)
      {
        return std::size_t(&*h) /
          sizeof( typename std::iterator_traits<H>::value_type);
      }
`
The iterator is default-constructed, holding a nullptr. operator*() creates a null reference by de-referencing the nullptr, which is undefined behaviour, even if the null reference is never accessed (as in this case).

Sorry I don't know enough about the internals of the library to know the best way to fix it.

This was found with UBsan's -fsanitize=null option ("Use of a null pointer or creation of a null reference").

Reference: N4861 (https://timsong-cpp.github.io/cppwp/n4861/dcl.ref)
9.3.3.2:
> A reference shall be initialized to refer to a valid object or function. [ Note: In particular, a null reference cannot exist in a well-defined program, because the only way to create such a reference would be to bind it to the “object” obtained by indirection through a null pointer, which causes undefined behavior. As described in [class.bit], a reference cannot be bound directly to a bit-field. — end note ]

## Source Code

This is called, for example, in the implicit copy constructor of Polyhedron_3.

`
Polyhedron_3 a;
Polyhedron_3 b(a);
`
(run with the UBsan option above to repro)

## Environment

* Compiler: clang++
* Specific flags used (if any): -fsanitize=null
* CGAL version: d9b0e84d8d4d1299e14a8bdc78c2ad414552640c

**Comment by tkoeppe** [comment]
The error seems to originate here: https://github.com/CGAL/cgal/blob/e3d5fccf5b9514f1c0ddcca4d417a011b5b26a9e/HalfedgeDS/include/CGAL/HalfedgeDS_list.h#L345

Here we're attempting to insert a default-constructed iterator into the map, but the map's hasher only supports dereferenceable iterators (and indeed the hash value is just the address of the pointed-to element):

https://github.com/CGAL/cgal/blob/a8e87526348d2d1faa2ea893747314853ee6868d/Hash_map/include/CGAL/Handle_hash_function.h#L36

(Note here that `*h` is assumed to be valid.)

**Comment by lrineau** [comment]
I think `std::size_t(&*h)` should be replaced by `h.operator->()`.

**Comment by afabri** [comment]
@GilesBathgate this seems like a deja-vu.  Did we discuss this in your PRs about hashing?

**Comment by GilesBathgate** [comment]
@afabri No, but I am familiar with this code, I will take a look and give an update.

**Comment by tkoeppe** [comment]
Thank you, everyone!

**Comment by GilesBathgate** [comment]
This program demonstrates the issue:

```
#include <iostream>

struct Handle
{
    int* ptr;
    Handle() : ptr(0) {}
    
    int& operator*() { return *ptr; }
    int* operator->() { return ptr; }
};

int main()
{
    Handle h;
    auto n1 = &*h; //UB
    std::cout << n1 << std::endl; 
    
    auto n2 = h.operator->();
    std::cout << n2 << std::endl;
    
    auto n3 = *h;
    std::cout << n3 << std::endl;
}
```

As you can see `&*h` doesn't segfault, but `*h` does. I think the UBSan is getting confused. Either way @lrineau 's suggestion seems to be a good solution.

**Comment by tkoeppe** [comment]
Hm, I'm not exactly sure which iterator type is being used here, but at least some of the iterators look like their default-constructed (and indeed their value-initialized) states allow calling `operator->()` (e.g. the filtered iterator forwards `&*` to its stored base iterator, which may be uninitialized (and will definitely be not dereferenceable) at that point).

**Comment by tkoeppe** [comment]
@GilesBathgate: I think your code is not valid C++. For `&*e` to be valid, `*e` must be valid,

(This is different in other programming languages, e.g. in the C language, `&*` is defined as a no-op regardless of whether `*` is valid. But that's a completely different language.)

**Comment by tkoeppe** [comment]
I haven't thought about this very hard, but I have a feeling that something like the following will be necessary: If you want iterators to be hashable, you need to decide if that includes a) past-the-end iterators, and b) singular iterators (i.e. iterators not associated with any sequence, for example, uninitialized pointers). If you do, then you need to find a way to make the those non-dereferenceable iterators detectable by the hash function.

A typical way to do this would be to make past-the-end iterators identical to singular ones, which would then be the value of a default-constructed object. (This is what the istream iterators do.) That way the hasher could detect that an iterator is that special kind just by looking at it in isolation, without reference to any range or container.

But that'd be quite a significant design change of your iterator types.

An alternative would be to reconsider whether you need the iterator in the map at all, or if you could maybe put some other (optional?) type in the map.

**Comment by GilesBathgate** [comment]
@tkoeppe I'm just documenting how the code behaves, vs how it's defined to behave.

**Comment by GilesBathgate** [comment]
>If you want iterators to be hashable

These are typically just Handles, not Iterators. but I think they are sometimes unintentionally mixed up in the codebase in some places since they are interchangeable.

https://doc.cgal.org/latest/Circulator/index.html#Chapter_Handles_Ranges_and_Circulators

**Comment by lrineau** [comment]
For our home-made iterators, `iterator.operator->` is always valid, but the returned pointer might not be dereferencable.

**Comment by tkoeppe** [comment]
> @tkoeppe I'm just documenting how the code behaves, vs how it's defined to behave.

@GilesBathgate: I see, yes. I just wanted to clarify that the problem is really with the C++ code, not with the sanitizer. The sanitizer is merely pointing the problem out.

**Comment by tkoeppe** [comment]
@lrineau: Ah, OK. I wasn't sure which iterators are used here. I only looked around some of the names (which I suppose are reused a lot) and saw this iterator, which dereferences inside `operator->`:
 https://github.com/CGAL/cgal/blob/a8e87526348d2d1faa2ea893747314853ee6868d/Arrangement_on_surface_2/include/CGAL/Arrangement_2/Arrangement_2_iterators.h#L86

**Comment by GilesBathgate** [comment]
@lrineau Things like this cause issues with `h.operator->()`

https://github.com/CGAL/cgal/blob/a8e87526348d2d1faa2ea893747314853ee6868d/Nef_2/include/CGAL/Nef_2/Segment_overlay_traits.h#L836-L837

**Comment by GilesBathgate** [comment]
@tkoeppe Likewise: https://github.com/CGAL/cgal/blob/a8e87526348d2d1faa2ea893747314853ee6868d/STL_Extension/include/CGAL/Multiset.h#L399-L404

**Comment by afabri** [comment]
I must miss something important here.  The iterator is the result of an insertion, so how can it be null?

**Comment by GilesBathgate** [comment]
@afabri The iterator isn't null, the node inside the iterator is null.

**Comment by GilesBathgate** [comment]
@afabri Ah, so what I was saying about `x2iso[&*it1]` leads to `std::size_t(&*&*it1)` but with @lrineau 's suggestion it would be `(&*it1).operator->()` which gives the compile error: did you mean to use ->

**Comment by lrineau** [comment]
> @afabri Ah, so what I was saying about `x2iso[&*it1]` leads to `std::size_t(&*&*it1)` but with @lrineau 's suggestion it would be `(&*it1).operator->()` which gives the compile error: did you mean to use ->

It should be `x2iso[it1.iterator->()]`.

**Comment by GilesBathgate** [comment]
@lrineau I can't get that to compile.

**Comment by matthewlai** [comment]
It sounds like the fundamental issue is whether operator*() and operator->() are supposed to be valid for empty iterators, and different iterators were written under different assumptions.

Maybe iterators can have an "operator bool()" that checks whether the iterator points to a valid object in whatever way makes sense for the iterator, and the pre-condition for operator*() and operator->() in the interface to require the iterator to be valid?

**Comment by GilesBathgate** [comment]
@matthewlai That would also work with `operator==(nullptr_t)` which I added to some iterators recently. (operator bool is a bit more generalised which may be better or not, it is how STL does it though)

**Comment by matthewlai** [comment]
@GilesBathgate To me operator bool is a bit more readable because the convention in C++ is that bool conversion signifies checking whether something is valid or not (works for std iterators, std streams, pointers, return code for C-style functions, etc). With "std::size_t(h === nullptr ? 0 : &*h)" people unfamiliar with the codebase may assume that h is a pointer, or convertible to a pointer.

**Comment by tkoeppe** [comment]
@matthewlai: It would be a major redesign of the iterator concept to make iterators "fat" by remembering their range; you're essentially proposing to replace every iterator by an iterator pair. While that's possible in principle, it represents a major change of direction of this core part of the C++ library conventions, and has non-trivial performance trade-offs. (E.g. users other that this present discussion would not need or want the proposed 100% size increase.)

Note that it dereferenceabality is already a predcondition for `*it`, and `it->x` is defined as `(*it).x`. (Iterators don't need to be class types, and thus don't need to have overloaded operators.) It's just that the precondition isn't _checkable_, since that would require the above-mentioned performance trade-offs. (A.k.a. "Java- style iterators").

**Comment by GilesBathgate** [comment]
>It would be a major redesign of the iterator concept to make iterators "fat" by remembering their range;
@tkoeppe He didn't say or imply that at all.

**Comment by GilesBathgate** [comment]
> To me operator bool is a bit more readable because the convention in C++ is that bool conversion signifies checking whether something is valid

@matthewlai I agree in hindsight, I just wanted to reference the existing change: https://github.com/CGAL/cgal/pull/5665

**Comment by matthewlai** [comment]
@tkoeppe Why would they need to remember their range? The way I see it, there are already ways to check whether each iterator type is empty, and there's just no interface for that.

**Comment by GilesBathgate** [comment]
would `std::addressof(h)` work?

**Comment by tkoeppe** [comment]
@matthewlai: it depends on whether you make it a design point that every possible non-dereferenceable state (that is, singular states as well as past-the-end states) is that specific null state. But then how do you implement `++it`? How do you know when to switch over to the null state?

**Comment by matthewlai** [comment]
@tkoeppe That is true. I was under the assumption that we only need one sentinel value, like if the iterator was just a naked pointer.

**Comment by GilesBathgate** [comment]
@tkoeppe @matthewlai Bear in mind we are discussing `Handle_hash_function`, not a supposed `Iterator_hash_function`

**Comment by tkoeppe** [comment]
@GilesBathgate: Yes, good point, you may very well be able to find a concise local solution for the particular use case.

**Comment by matthewlai** [comment]
> would `std::addressof(h)` work?

That would take the stack address of the iterator I think.

**Comment by GilesBathgate** [comment]
Maybe iterators/handles should implement `size_t operator&() { return size_t(ptr); }` since I can't think the adress of the handle or iterator is that useful or needed, (if it is needed one can always use `std::addressof`) The hash functor then just becomes `&h / sizeof(...)`

**Comment by matthewlai** [comment]
Personally I would recommend against it, since it's a very rarely overloaded operator (in fact, I didn't know it's overloadable until today), and it can make the calling code difficult to follow for people who don't know about this oddity in the codebase. If I read some code that says '&x', I am not thinking in my head that operator& may be overloaded and doing something else entirely.

I personally think operator& overloads are very dangerous, because it's already implemented for all types, and it actually means the same thing for all types (without an overload). This is unlike operator* for example, where it's only defined for pointers, so you know if it's used with a custom type, it must be an overload, and it also has a well known semantic (it's "the object this thing refers to", whether that's a pointer, a smart pointer, an iterator, or an optional). operator&'s logical meaning is "address of". Address of an iterator on the stack being actually the address of the object pointed to by the iterator is not obvious.

What advantage does it have over bool conversion?

**Comment by GilesBathgate** [comment]
@matthewlai I was rather looking for an alternative to `h.operator->()` because that has issues with the `.` as discussed above. 
`size_t(h ? &*h : 0)` fixes the sanitiser issue, the only disadvantage I can think of is additional asm code generation, although maybe it can be optimised away.

**Comment by GilesBathgate** [comment]
@matthewlai Actually you can also just apply your fix to the handle class?

`T& operator*() { return node ? *node : 0; }`

**Comment by matthewlai** [comment]
Pretty sure it will be optimised away. The operator bool call will almost certainly be inlined, and any reasonable compiler will then optimise out the branch.

>> `T& operator*() { return node ? *node : 0; }`
This doesn't work because it would still create a null reference (also I don't think you can create a l-value reference from an r-value).

**Comment by mglisse** [comment]
In C++20, there is `std::to_address`.

**Comment by GilesBathgate** [comment]
@mglisse Ah very good. You could create a CGAL implementation of that based on the possible implementation here https://en.cppreference.com/w/cpp/memory/to_address But it's in essence the same as what was implemented in #5731
